mirrord/layer/tests/outgoing.rs

```diff
@@ -0,0 +1,135 @@
+#![feature(assert_matches)]
+#![cfg(target_os = "linux")]
+
+use std::{net::SocketAddr, path::PathBuf, time::Duration};
+
+use mirrord_protocol::{
+    outgoing::{
+        tcp::{DaemonTcpOutgoing, LayerTcpOutgoing},
+        udp::{DaemonUdpOutgoing, LayerUdpOutgoing},
+        DaemonConnect, DaemonRead, LayerConnect, LayerWrite, SocketAddress,
+    },
+    ClientMessage, DaemonMessage,
+};
+use rstest::rstest;
+
+mod common;
+
+pub use common::*;
+use futures::{SinkExt, TryStreamExt};
+
+/// Test outgoing UDP.
+/// Application, for each remote peer in [`RUST_OUTGOING_PEERS`]:
+/// 1. Opens a UDP port at [`RUST_OUTGOING_LOCAL`]
+/// 2. Connects to the remote peer
+/// 3. Sends some data
+/// 4. Expects the peer to send the same data back
+///
+/// # Ignored
+/// This test is ignored due to a bug - `recv_from` call returns an invalid remote peer address.
+#[ignore]
+#[rstest]
+#[tokio::test]
+#[timeout(Duration::from_secs(10))]
+async fn outgoing_udp(dylib_path: &PathBuf) {
+    let (mut test_process, layer_connection) = Application::RustOutgoingUdp
+        .start_process_with_layer(dylib_path, vec![], None)
+        .await;
+    let mut conn = layer_connection.codec;
+
+    let peers = RUST_OUTGOING_PEERS
+        .split(',')
+        .map(|s| s.parse::<SocketAddr>().unwrap())
+        .collect::<Vec<_>>();
+
+    for peer in peers {
+        let msg = conn.try_next().await.unwrap().unwrap();
+        let ClientMessage::UdpOutgoing(LayerUdpOutgoing::Connect(LayerConnect { remote_address: SocketAddress::Ip(addr) })) = msg else {
+            panic!("Invalid message received from layer: {msg:?}");
```

Comment:
Why are you panicking rather than asserting (here and elsewhere)?

Reply:
The `assert_matches` macro we have does not allow for binding values. For example, in line 62 I bind the `bytes` field, which is embedded deeply into the message.